lib/ws-kernel-picker.js

```diff
@@ -73,67 +77,80 @@ export default class WSKernelPicker {
-            });
-            return sessionListing.setItems(items);
-          },
-          () =>
-            // Gateways offer the option of never listing sessions, for security
-            // reasons.
-            // Assume this is the case and proceed to creating a new session.
-            this.onSession({
-              name: "[new session]",
-              model: null,
-              options: gatewayInfo.options,
-              kernelSpecs
-            })
-        );
-      },
-      () => atom.notifications.addError("Connection to gateway failed")
+    const sessionListing = new CustomListView(
+      "No sessions available",
+      this.onSession.bind(this)
     );
+    services.Kernel
+      .getSpecs(gatewayInfo.options)
+      .then(
+        specModels => {
+          const kernelSpecs = _.filter(specModels.kernelspecs, spec =>
+            this._kernelSpecFilter(spec)
+          );
+
+          const kernelNames = _.map(kernelSpecs, specModel => specModel.name);
+
+          sessionListing.previouslyFocusedElement = this.previouslyFocusedElement;
+          sessionListing.selectListView.update({
+            loadingMessage: "Loading sessions..."
+          });
+
+          services.Session.listRunning(gatewayInfo.options).then(
+            sessionModels => {
+              sessionModels = sessionModels.filter(model => {
+                const name = model.kernel ? model.kernel.name : null;
+                return name ? kernelNames.includes(name) : true;
+              });
+              const items = sessionModels.map(model => {
+                let name;
+                if (model.notebook && model.notebook.path) {
+                  name = tildify(model.notebook.path);
+                } else {
+                  name = `Session ${model.id}`;
+                }
+                return {
+                  name,
+                  model,
+                  options: gatewayInfo.options
+                };
+              });
+              items.unshift({
+                name: "[new session]",
+                model: null,
+                options: gatewayInfo.options,
+                kernelSpecs
+              });
+              return sessionListing.selectListView.update({
+                items: items,
+                loadingMessage: null
+              });
+            },
+            () =>
+              // Gateways offer the option of never listing sessions, for security
+              // reasons.
+              // Assume this is the case and proceed to creating a new session.
+              this.onSession({
+                name: "[new session]",
+                model: null,
+                options: gatewayInfo.options,
+                kernelSpecs
+              })
+          );
+        },
+        () => atom.notifications.addError("Connection to gateway failed")
+      )
+      .then(() => {
+        sessionListing.selectListView.focus();
+        sessionListing.selectListView.reset();
+      });
   }
```

Comment:
Lines 150-153 above are pretty much all the recent commits add, the rest is just prettier formatting.